src/psyclone/psyir/nodes/array_mixin.py

```diff
@@ -188,6 +188,35 @@ def is_lower_bound(self, index):
             return False
         return True
 
+    def lbound(self, pos):
+        '''
+        Lookup the lower bound of the specified dimension of this ArrayMixin.
+        If we don't have the necessary type information then a call to the
+        LBOUND intrinsic is constructed and returned.
+
+        :param int pos: the dimension of the array for which to lookup the \
+                        bounds.
+
+        :returns: the declared lower bound for the specified dimension of \
+            this ArrayMixin or a call to the LBOUND intrinsic if it is not \
+            known.
+        :rtype: :py:class:`psyclone.psyir.nodes.Node`
+
+        '''
+        if not hasattr(self, "symbol"):
+            # If we are a Member of some sort then we need to carry on up the
+            # Method Resolution Order list to call the appropriate version
+            # of this method.
+            return super().lbound(pos)
```

Comment:
The problem is that we can have a `Member` that actually represents an access to an entire array within a structure. Without fully resolving the type information of the parent structure, we can't do any better than this. I'm not sure what we do about that?

Reply:
Aha, I see. Because it is accessors without the array index e.g. `call sub(my_struct%my_array)` and we need the lbound of this expression? But don't we have the same problem with a single reference? e.g.
```
real, dimension(5:10) :: array
call subroutine(array)
```
Reference does not have `.lbound`? Do we do something different or would this currently produce wrong inlined code?

Reply:
I tried the inline transformation with:
```
         "module test_mod\n"
            "contains\n"
            "  subroutine run_it()\n"
            "  real :: a\n"
            "  real, dimension(5:10) :: b\n"
            "  call sub(b)\n"
            "  end subroutine run_it\n"
            "  subroutine sub(b)\n"
            "    real, dimension(:) :: b\n"
            "    b(1) = 5\n"
            "  end subroutine\n"
            "end module test_mod\n")
```
and returned:
```
  subroutine run_it()
    real :: a
    real, dimension(5:10) :: b
    b(1) = 5
  end subroutine run_it
```
which is wrong

Reply:
In fact the "_update_actual_indices" utility in inline transformation is only called in line 492 as:
```
if hasattr(cursor, "indices"):
    new_indices = self._update_actual_indices(
```
So currently it will never call the Member.lbound from outside the ArrayMixin.

I see one possible solution:
In the inline transformation check if references and members (not subclasses of these) are of datatype ArrayType and in this case replace them with the expressions.
`array` -> `array(:)`
`struct%array` -> `struct%array(:)`
Forbid inlining references/member (not subclasses of these) that are still DeferredType.
Then continue with the code that you already have, and you can now move the Member.lbound inside ArrayMixin.

Would that work?

Reply:
I can use `Reference2ArrayRangeTrans` to do the conversion of the actual arguments (although it doesn't yet support arrays within structures - #1858 - so this won't work for GOcean kernel calls!).

Reply:
Actually, in GOcean kernels we don't specify the bounds of the dummy arguments so I've extended the code to check for that - if the bounds aren't specified locally then we don't have to worry as they are then the same as at the call site.